catkin_ws/src/state_estimation/src/state_aggregator.py

```diff
@@ -4,358 +4,103 @@
-        orientation = Quaternion(x=q_auv_world.x, y=q_auv_world.y, z=q_auv_world.z, w=q_auv_world.w)
-        pose = Pose(position, orientation)
-        self.pub_auv.publish(pose)
-
-        # publish individual degrees of freedom
-        self.pub_x.publish(pos_auv_world[0])
-        self.pub_y.publish(pos_auv_world[1])
-        self.pub_z.publish(pos_auv_world[2])
-
-        # publish world pose (relative to global)
-        position = Point(x=self.pos_world_global[0], y=self.pos_world_global[1], z=self.pos_world_global[2]) 
-        orientation = Quaternion(x=self.q_world_global.x, y=self.q_world_global.y, z=self.q_world_global.z, w=self.q_world_global.w)
-        pose = Pose(position, orientation)
-        self.pub_world.publish(pose)
-
-        # backwards compatibility
-        self.pub_theta_x.publish(euler_auv_world[0])
-        self.pub_theta_y.publish(euler_auv_world[1])
-        self.pub_theta_z.publish(euler_auv_world[2])
-
-        # angular velocity - quaternion PID
-        augular_velocity = Vector3(w_auv[0], w_auv[1], w_auv[2])
-        self.pub_w_auv.publish(augular_velocity)
-
+from tf2_ros import TransformBroadcaster
+
+from geometry_msgs.msg import Pose, Quaternion, Vector3, TransformStamped, Point
+from sensors import DepthSensor, IMU, DVL
+from std_msgs.msg import Float64
+
+DEG_PER_RAD = 180 / np.pi
+
+def update_state(_):    
+    x = None
+    y = None
+    z = None
+    quaternion = None
+
+    for sensor in sensor_priorities["x"]:
+        if sensor.isActive():
+            x = sensor.x
+            break
+    for sensor in sensor_priorities["y"]:
+        if sensor.isActive():
+            y = sensor.y
+            break
+    for sensor in sensor_priorities["z"]:
+        if sensor.isActive():
+            z = sensor.z
+            break
+    for sensor in sensor_priorities["orientation"]:
+        if sensor.isActive():
+            np_quaternion = np.array([sensor.q_nwu_auv.x, sensor.q_nwu_auv.y, sensor.q_nwu_auv.z, sensor.q_nwu_auv.w])
+            quaternion = Quaternion(x = sensor.q_nwu_auv.x, y = sensor.q_nwu_auv.y, z = sensor.q_nwu_auv.z, w = sensor.q_nwu_auv.w)
+            roll = transformations.euler_from_quaternion(np_quaternion, 'rxyz')[0] * DEG_PER_RAD
+            pitch = transformations.euler_from_quaternion(np_quaternion, 'ryxz')[0] * DEG_PER_RAD
+            yaw = transformations.euler_from_quaternion(np_quaternion, 'rzyx')[0] * DEG_PER_RAD
+            angular_velocity = Vector3(sensor.angular_velocity[0], sensor.angular_velocity[1], sensor.angular_velocity[2])
+            break
+
+    if x is not None and y is not None and z is not None and quaternion is not None:
+        pub_x.publish(x)
+        pub_y.publish(y)
+        pub_z.publish(z)
+        pub_theta_x.publish(roll)
+        pub_theta_y.publish(pitch)
+        pub_theta_z.publish(yaw)
+        pub_ang_vel.publish(angular_velocity)
+        pose = Pose(Point(x=x, y=y, z=z), quaternion)
```

Comment:
we should probably change this so that we can still have a quaternion without an x or y position, or vice-versa (just cuz im thinking of times when the DVL has broken or IMU and we still wanted to rotate or whatever)

Reply:
this is kind of a big rewrite we should make it its own issue

Reply:
it affects what the controls subscribes to and I think the planner as well

Reply:
True I hadn't though of that, I'll make the new issue now